### ISSUE TITLE
Over-strict keywords prevent HQ cover matching from google source

### ISSUE THREAD

**Issue by baoshan** [desbma/sacad#10]
This may be a too specific case - not general enough to make us any decision. I'm reporting here for future reference:

`sacad -t 100 "Claudio Arrau" "Arrau Heritage - Bach" 1600 cover.jpg` (album name from gracenote)
`sacad -t 100 "Claudio Arrau" "Arrau Heritage: Bach" 1600 cover.jpg` (album name from musicbrainz)

Above commands only offer a scanned version (from AmazonCD Source)

But if we loosen the google source keywords from `"Claudio Arrau" "Arrau Heritage - Bach" front cover` to `Claudio Arrau Arrau Heritage - Bach`, the __first match__ is an authentic digital cover of size `1400x1400` (from Amazon.co.jp).

Please advice if you think we could do something.

Thanks very much.

**Comment by desbma** [comment]
Which SACAD version are you using?

I'm getting [this cover](http://www.music-bazaar.com/album-images/vol1011/402/402022/2236396-big/Chopin-Arrau-Heritage-CD-8-cover.jpg) (1944x1944, from Google Images) for both you examples.

**Comment by baoshan** [comment]
I'm using `1.6.7`, cache flushed.

```
sacad -t 100 "Claudio Arrau" "Arrau Heritage - Bach" 1600 cover.jpg
Search with source 'LastFmCoverSource' failed: HTTPError 400 Client Error: Bad Request for url: https://ws.audioscrobbler.com/2.0/?method=album.getinfo&api_key=2410a53db5c7490d0f50c100a020f359&album=arrau+heritage+-+bach&artist=claudio+arrau
Got 2 relevant (0 excluded) results from source 'AmazonCdCoverSource'
Got 0 relevant (0 excluded) results from source 'AmazonDigitalCoverSource'
Got 0 relevant (0 excluded) results from source 'GoogleImagesWebScrapeCoverSource'
No reference result found
Downloading cover 'https://images-na.ssl-images-amazon.com/images/I/51D2B0896BL.jpg' (part 1/1)...
Crunching JPEG image...
```

**Comment by baoshan** [comment]
The different results from Google Image source is because I'm issuing the request from Tokyo, Japan (VPN). I can also get your aforementioned picture when connected to a US VPN.

If I loosen the keywords to `"%s %s" % (artist.lower(), album.lower())` and limit the result to the first entry (quick and dirty, sorry). I get [this image](http://ecx.images-amazon.com/images/I/71kn-G13l4L._AA1400_.jpg).

**Comment by desbma** [comment]
I see 3 different things about this issue:

1. About removing the "front cover" keywords at the end of the Google query: I will not do it because the amount of unrelated image will skyrocket. In your example it may work better without it because there is no ambiguity about what is being searched, but because Google Images is not specialized in covers, for other query this could lead to torally unrelated results. Since it was the main point, I'm closing this issue.

2. About the punctuation in the query: if you notice cases where results are different (and better) without it, please open another issue. If that is the case, that should be easy to fix.

3. I'm a bit surprised that Google changes the results only from the geographic origin of the IP, I thought they did that from the `Accept-Language` HTTP header or country TLD. I'd like to disable that if possible to get country neutral results, but that is another issue.

**Comment by baoshan** [comment]
>  I will not do it because the amount of unrelated image will skyrocket.

It's a dilemma. It's actually balancing between false-positive and false-negative. By add the `front cover`, There's a big chance we miss a lot of pages who didn't mentioned `front cover` in the text (or the metadata) in favor of accuracy.

I also favor accuracy over match rate. I can see a possible solution here:

+ I noticed all sources are ordered and queried in order. Before we request the Google Image source, we have reference images already. Can we leverage the know of those image to:

  - Remove the `front cover` keywords, and the `"`, then discard those images beyond a tolerance of similarity.
  - Use the reference images to submit a Google similarity image search, then discard those images beyond a tolerance of similarity. (Google may offer images with a similarity in semantics, not pixel).

> I'm a bit surprised that Google changes the results only from the geographic origin of the IP,

An `API` may be more predictable. What `SACAD` is doing is web page analysising.

**Comment by baoshan** [comment]
Also, we may replace `front cover` for an aspect ratio between `0.95 to 1.05` (is there one?). Which requires a set of `artist / album` keywords to benchmark the performance to see if there's a gain.

**Comment by desbma** [comment]
>  I noticed all sources are ordered and queried in order. [...]

That won't be the case for ever, I have in my TODO list do do that in parallel threads (potential huge performance gain).

Google Images is typically only useful for rare covers not found on other sources, so trying to improve results from the reference is not worth it in my opinion, because when we really need Google Images, there is no reference cover.

> An API may be more predictable. What SACAD is doing is web page analysising.

A few years ago (before SACAD was open source, or even named SACAD), I was using the Google Images JS API, but they were rate limiting aggressively, and the API is gone now anyway.
